libnavui-androidauto/src/main/java/com/mapbox/androidauto/car/placeslistonmap/PlacesListOnMapManager.kt

```diff
@@ -0,0 +1,140 @@
+    private val placesListOnMapProvider: PlacesListOnMapProvider,
+) : MapboxCarMapObserver {
+
+    private var carMapSurface: MapboxCarMapSurface? = null
+    private var coroutineScope: CoroutineScope? = null
+    private var styleLoadedListener: OnStyleLoadedListener? = null
+    private var placesListItemMapper: PlacesListItemMapper? = null
+    private val placesLayerUtil: PlacesListOnMapLayerUtil = PlacesListOnMapLayerUtil()
+    private val navigationObserver = mapboxNavigationForward(this::onAttached) { onDetached() }
+
+    private val _placeRecords = MutableStateFlow(listOf<PlaceRecord>())
+    val placeRecords: StateFlow<List<PlaceRecord>> = _placeRecords.asStateFlow()
+
+    private val _placeSelected = MutableStateFlow<PlaceRecord?>(null)
+    val placeSelected: StateFlow<PlaceRecord?> = _placeSelected.asStateFlow()
+
+    private val placeClickListener = object : PlacesListItemClickListener {
+        override fun onItemClick(placeRecord: PlaceRecord) {
+            logAndroidAuto("PlacesListOnMapScreen request $placeRecord")
+            _placeSelected.value = placeRecord
+        }
+    }
+
+    fun currentItemList(): ItemList? {
+        val carAppLocation = MapboxNavigationApp.getObserver(CarAppLocation::class)
+        val currentLocation = carAppLocation.navigationLocationProvider.lastLocation
+            ?: return null
+        return placesListItemMapper?.mapToItemList(
+            currentLocation,
+            placeRecords.value,
+            placeClickListener
+        )
+    }
+
+    override fun onAttached(mapboxCarMapSurface: MapboxCarMapSurface) {
+        super.onAttached(mapboxCarMapSurface)
+        carMapSurface = mapboxCarMapSurface
+        coroutineScope = MainScope()
+        MapboxNavigationApp.registerObserver(navigationObserver)
+
+        styleLoadedListener = mapboxCarMapSurface.handleStyleOnAttached {
+            placesLayerUtil.initializePlacesListOnMapLayer(
+                it,
+                mapboxCarMapSurface.carContext.resources
+            )
+            loadPlaceRecords()
+        }
+    }
+
+    override fun onDetached(mapboxCarMapSurface: MapboxCarMapSurface) {
+        super.onDetached(mapboxCarMapSurface)
+        mapboxCarMapSurface.handleStyleOnDetached(styleLoadedListener)?.let {
+            placesLayerUtil.removePlacesListOnMapLayer(it)
+        }
+        styleLoadedListener = null
+        MapboxNavigationApp.unregisterObserver(navigationObserver)
+        carMapSurface = null
+    }
+
+    private fun onAttached(mapboxNavigation: MapboxNavigation) {
+        placesListItemMapper = PlacesListItemMapper(
+            PlaceMarkerRenderer(carMapSurface?.carContext!!),
+            mapboxNavigation
+                .navigationOptions
+                .distanceFormatterOptions
+                .unitType
+        )
+    }
+
+    private fun onDetached() {
+        placesListItemMapper = null
+        coroutineScope?.cancel()
+        coroutineScope = null
```

Comment:
Should this be in `onDetached(MapboxCarMapSurface)`, since the scope is created in `onAttached(MapboxCarMapSurface)`?

Reply:
Definitely!

Reply:
for reference https://github.com/mapbox/mapbox-navigation-android/pull/6398